SIMD-0106: Epoch Rewards Partition Data PDA (#106)
solana-foundation/solana-improvement-documents: proposals/0106-paritioned-epoch-rewards-pda.md

---
simd: '0106'
title: Epoch Rewards Partition Data PDA
authors:
  - Tyera Eulberg
category: Standard
type: Core
status: Draft
created: 2024-01-17
feature: https://github.com/solana-labs/solana/issues/32166
extends: 0015
---

## Summary

Partitioned epoch rewards (SIMD 0015) will split epoch rewards distribution
across multiple blocks at the start of an epoch. This extension of that SIMD
proposes storing data about the rewards calculation and partitions in a sysvar
account at a program-derived address (PDA), which can be queried by clients, or
indeed within the runtime, to performantly identify the partition for the
distribution to any particular address.

> ## Motivation

Reviewer: Can you also include the motivation for including this data on-chain? I personally think that storing it in the first blocks metadata would be great. We already use the metadata to record the list of rewards.

Author: Added. But yes, planning to do both. Will write up a separate SIMD for the metadata approach.

Author: I reverted this one as well. Here's what I had written:
"If the partition data were stored in an on-chain account, it would be simple and
low-compute-cost for a client (with a snapshot or running node) or the runtime
itself to find the rewards for a particular address for any historical epoch
after partitioned epoch rewards are activated."

I do feel like this is already articulated in the summary.

Author: > Ok, how open are you to removing the PDA approach and just using the metadata?

Somewhat opposed. I think it makes sense to have some sort of persistent record of the partitioned-rewards data on chain, but it's true I don't know of a specific in-program use-case for it. From the rpc perspective, it's also really nice to be able to figure out the rewards block before having to determine what data store to use to find that block.
@t-nelson , perhaps you'd care to chime in here?

Reviewer: Is that very different from the rpc fetching the first block in the epoch to find out where the rewards are? If I recall correctly, there's an extra call getBlocks to figure out what the first block actually is, is that the part you're trying to improve?

Author: > there's an extra call getBlocks to figure out what the first block actually is, is that the part you're trying to improve?

Yes. There's potentially quite a bit of latency in each long-term storage query, so it would be nice to avoid the extra call.

Reviewer: If the main motivation is to save some latency overhead on step 1 of fetching rewards when step 2 requires high latency big table fetching most of the time anyways, it doesn't seem like this proposal is worth the runtime complexity.

When we move to partitioned epoch rewards as described in the original SIMD, the
only way to find the stake or voting rewards for a specific address will be to
iterate through blocks at the beginning of each epoch, inspecting all the
rewards. This is because the runtime does not persist information about how the
rewards were partitioned; in fact, it does not even persist how many blocks the
rewards distribution spans, so there is no way to predict how long it will take
or how far through the epoch to go in finding a particular address.

## Alternatives Considered

An alternative to storing partition data in an on-chain account would be to
record the necessary data in the ledger in some fashion. This could be a
transaction that gets added to the block, a special RewardType, or metadata that
gets stored on the node and then duplicated to long-term storage, like Bigtable.
In fact, it is probably worthwhile to pursue this alternative as well, since it
will enable finding rewards without access to snapshots or the running chain.

## New Terminology

None

## Detailed Design

> When partitioned rewards are calculated in the runtime (currently during the

Reviewer: We should be more specific about when in the first block this account is created.

Author: I can add that, but I feel that's really something that should be defined in the [original partitioned-rewards SIMD](https://github.com/solana-foundation/solana-improvement-documents/blob/main/proposals/0015-partitioned-epoch-reward-distribution.md), if it needs to be specified in a SIMD. I feel it is sufficient and correct for this SIMD to say that the sysvar population happens as part of the Rewards Calculation operation (whenever that might be).

Reviewer: I see, I was specifically just concerned that the combination of the two SIMD's make it clear that the PDA should be created by the runtime before transactions are processed. Some of the wording had felt a little loose and ambiguous. I think your change makes it more clear, thanks!

Reviewer: I just read SIMD 15 and it specifically says that the sysvar is created at the beginning of the first block before processing transactions. Nitpicky but I think it would be nice if we are similarly clear about that here and say that this PDA sysvar is created at the same time as that one.

first block of the epoch), the runtime should populate a PDA that stores the
partition data needed to recreate the hasher that returns the partition index
for any address. This data comprises: the number of partitions, the parent
blockhash, and the hasher kind. More specifically:

```rust
// Version wrapper to allow future updates
// Variant serialized as little-endian unsigned 32-bit integer
enum EpochRewardsPartitionDataVersion {
    V0(PartitionData),
}

// Extensible list of kinds of hashers used to generate partitions
// Mapping fn defines the HasherKind for each partition-data version
enum HasherKind {
    Sip13 = 0,
}

// Data about rewards partitions for a particular epoch
struct PartitionData {
    num_partitions: usize, // little-endian unsigned 64-bit integer
    parent_blockhash: Hash, // byte-array of length 32
}
```

The address of this PDA will use some bytes -- to prevent griefing and namespace
the PDAs -- and the rewards distribution epoch number as a little-endian u64 as
seeds. Specifically: `[b"EpochRewards",b"PartitionData", &epoch.to_le_bytes()]`.

> The owning program should be the sysvar program id:

Reviewer: Can you add the reasoning for why it's owned by the sysvar program rather than the stake program? And also comment on how these sysvars will be different from other sysvars? (Writable, for example)

Author: I don't actually know what you mean by "(Writable, for example)"

Author: I did take a crack at this, but both Trent and I thought it made the proposal worse, so I've reverted it. Here's what I had written:
"Ownership by the Stake program was also considered, but would require either
extending the Stake state enum definitions or for parsers to do special handling
based on data length. Meanwhile, this PDA has more in common with existing
Sysvar-owned accounts: like other sysvars, it is written and controlled only by
the runtime. The only difference from a traditional sysvar is the
program-derived instead of a static address."

Reviewer: Maybe worth noting that these sysvars won't be accessible via syscalls since other sysvars typically are assumed to be.

`Sysvar1111111111111111111111111111111111111`.

## Impact

The change in this proposal does increase the number of "forever" accounts that
validators must store by one per epoch. However, the PDAs will be owned by the
stake program, so could be adjusted or closed in the future by a feature-gated

> change to the Stake Program. Meanwhile, the change greatly improves the

Reviewer: Should be the sysvar program now

post-SIMD 0015 situation for clients trying to track stake or voting rewards,
since they can use the data in the PDA to pull the correct partition directly,
instead of scanning an unknown number of blocks.

## Security Considerations

> Like traditional sysvars, the new accounts should only be loadable as read-only.

Reviewer: The SIMD should explain how this is done since it's also important for consensus. Are they read-only only after they are created?

Author: As you stated on your SIMD, I think how dynamically addressed sysvars would have their write-locks demoted is outside the scope of this proposal. But I expanded the section a little.

Reviewer: This SIMD needs to address the specifics of write-lock demotion if you state that these sysvars should be loaded as read-only. Right now we only have a way to use feature gates to introduce new read-only sysvars.. we don't have a design for this new kind of sysvar which is created each epoch yet. So I think this SIMD can reference SIMD 105 and say that the new sysvars will be added to the dynamic list when they are created on-chain. Also, I don't think that the "Security considerations" is the right place to discuss read-only loading.. it should be in the design section.

Reviewer: I'm fine with handling in a separate SIMD but this SIMD says they should only be loadable as read-only but doesn't explain that the design is still TBD. So I think we should say that it's TBD or design it here.

Reviewer: As for lack of persistence, we can add all these sysvars in a loop from some starting epoch to the current epoch. This seems fine to do in each restart because epochs will only be in the hundreds to thousands.

Author: I feel like the use of "should" here is consistent with rfc2119, and saying "a new proposal should be introduced" is the same thing as saying the design is TBD. But I'm also happy to remove the paragraph entirely.

Reviewer: Oh sorry I didn't realize "should" meant that

If the partition data were stored in an on-chain account, it would be simple and
low-compute-cost for a client (with a snapshot or running node) or the runtime
itself to find the rewards for a particular address for any historical epoch
after partitioned epoch rewards are activated.

## Alternatives Considered

An alternative to storing partition data in an on-chain account would be to
record the necessary data in the ledger in some fashion. This could be a
transaction that gets added to the block, a special RewardType, or metadata that
gets stored on the node and then duplicated to long-term storage, like Bigtable.
In fact, it is probably worthwhile to pursue this alternative as well, since it
will enable finding rewards with just the historical ledger and no access to
snapshots or the running chain.

## New Terminology

None

## Detailed Design

When partitioned rewards are calculated in the runtime (currently in the first
block of the epoch when new-epoch operations -- like feature activations and
leader schedule generation -- are processed), the runtime should populate a PDA
that stores the partition data needed to recreate the hasher that returns the

> partition index for any address. This data comprises: the number of partitions,
> the parent blockhash, and the hasher kind. More specifically:

Reviewer 2: ```suggestion
partition index for any address. This data comprises: the number of partitions
and parent blockhash. More specifically:
```

hash function is now implied by the account data layout version

```rust
// Version wrapper to allow future updates
// Variant serialized as little-endian unsigned 32-bit integer
enum EpochRewardsPartitionDataVersion {
    V0(PartitionData),
}

> // Extensible list of kinds of hashers used to generate partitions
> // Mapping fn defines the HasherKind for each partition-data version
> enum HasherKind {
>     Sip13 = 0,
> }
> 

Reviewer 2: ```suggestion
```

i think? maybe we mention the hash function being SipHash13 as per SIMD0015 in v0 in the prose?

Author: Yeah, good idea. Unfortunately, SIMD 0015 didn't speak to the hasher at all (probably an update that someone should do), so see what you think about the way I've written it.

// Data about rewards partitions for a particular epoch
struct PartitionData {

>     num_partitions: usize, // little-endian unsigned 64-bit integer

Reviewer 2: ```suggestion
    num_partitions: u64, // little-endian unsigned 64-bit integer
```

target dependent types are target dependent

    parent_blockhash: Hash, // byte-array of length 32
}
```

The address of this PDA will use some bytes -- to prevent griefing and namespace
the PDAs -- and the rewards distribution epoch number as a little-endian u64 as
seeds. Specifically: `[b"EpochRewards",b"PartitionData", &epoch.to_le_bytes()]`.
The owning program should be the Sysvar program id:
`Sysvar1111111111111111111111111111111111111`.

Ownership by the Stake program was also considered, but would require either
extending the Stake state enum definitions or for parsers to do special handling
based on data length. Meanwhile, this PDA has more in common with existing
Sysvar-owned accounts: like other sysvars, it is written and controlled only by
the runtime. The only difference from a traditional sysvar is the
program-derived instead of a static address.

## Impact

The change in this proposal does increase the number of "forever" accounts that
validators must store by one per epoch. However, the PDAs will be owned by the

> stake program, so could be adjusted or closed in the future by a feature-gated

Reviewer 2: ```suggestion
Sysvar program, so could be adjusted or closed in the future by a feature-gated
```

rewards. This is because the runtime does not persist information about how the
rewards were partitioned; in fact, it does not even persist how many blocks the
rewards distribution spans, so there is no way to predict how long it will take
or how far through the epoch to go in finding a particular address.

## Alternatives Considered

An alternative to storing partition data in an on-chain account would be to
record the necessary data in the ledger in some fashion. This could be a
transaction that gets added to the block, a special RewardType, or metadata that
gets stored on the node and then duplicated to long-term storage, like Bigtable.
In fact, it is probably worthwhile to pursue this alternative as well, since it
will enable finding rewards without access to snapshots or the running chain.

## New Terminology

None

## Detailed Design

When partitioned rewards are calculated in the runtime (currently during the
first block of the epoch), the runtime should populate a PDA that stores the
partition data needed to recreate the hasher that returns the partition index
for any address. This data comprises: the number of partitions, the parent

> blockhash, and the hasher kind.

Reviewer 2: mind pseudocoding the data layout and hasher kind values?  kinda like [this](https://github.com/solana-foundation/solana-improvement-documents/blob/main/proposals/0064-transaction-receipt.md#transactionreceiptdata), but maybe make it smell less like rust :wink:

The address of this PDA should include the current epoch number (which contains

> the distributions) as a little-endian u64, as well as some bytes to prevent

Reviewer 2: the parenthetical here kinda throws me off. i'm not sure what it intends to add in this context?

Author: I was trying to distinguish between the epoch over which the rewards were earned (which is how rewards are queried) from the epoch in which the distributions were made (+1). I have removed the parenthetical, but please double-check that it's clear to you which epoch number is used in the PDA addressing.

> griefing (eg. `b"EpochRewardsPartitionData"`). The owning program should be the

Reviewer 2: it'd probably be more clear to present the seeds precisely in pseudocode (i think they're presented backwards here from the code?)

Stake Program.

## Impact

The change in this proposal does increase the number of "forever" accounts that
validators must store by one per epoch. However, the PDAs will be owned by the
stake program, so could be adjusted or closed in the future by a feature-gated
change to the Stake Program. Meanwhile, the change greatly improves the
post-SIMD 0015 situation for clients trying to track stake or voting rewards,
since they can use the data in the PDA to pull the correct partition directly,
instead of scanning an unknown number of blocks.

## Security Considerations

Like traditional sysvars, the new accounts should only be loadable as read-only.
How this would be accomplished depends on the outcome of SIMD 0105.

## Backwards Compatibility

> Population of an account each epoch is a consensus-breaking change and must be
> activated with a feature gate. Since this is an extension of SIMD 0015, it
> should be gated by the same feature id.

Reviewer 2: ```suggestion
Runtime-population of an account each epoch is a consensus-breaking change and
must be activated with a feature gate. Since this is an extension of SIMD 0015
and that feature gate has not yet been activated on any public clusters, it
should be gated by the same feature id.
```

record the necessary data in the ledger in some fashion. This could be a
transaction that gets added to the block, a special RewardType, or metadata that
gets stored on the node and then duplicated to long-term storage, like Bigtable.
In fact, it is probably worthwhile to pursue this alternative as well, since it
will enable finding rewards without access to snapshots or the running chain.

## New Terminology

None

## Detailed Design

When partitioned rewards are calculated in the runtime (currently during the
first block of the epoch), the runtime should populate a PDA that stores the
partition data needed to recreate the hasher that returns the partition index
for any address. This data comprises: the number of partitions, the parent
blockhash, and the hasher kind. More specifically:

```rust
// Version wrapper to allow future updates
enum EpochRewardsPartitionDataVersion {
    V0(PartitionData),
}

// Extensible list of kinds of hashers used to generate partitions
enum HasherKind {
    Sip13,
}

// Data about rewards partitions for a particular epoch
struct PartitionData {
    num_partitions: usize,
    parent_blockhash: Hash,
    hasher_kind: HasherKind,
}

> ```

Reviewer 2: thanks! from an implementer's perspective, we need to be more precise here. the enum layouts need to be specified (`8bit unsigned integer` for both?), `usize` is target-dependent, so we need to specify it as something like `64bit unsigned little-endian integer`. similarly replace `Hash` with something like `byte-array of length 32`

The address of this PDA should include some bytes to prevent griefing (eg.

> `b"EpochRewardsPartitionData"`) and the current epoch number as a little-endian

Reviewer 2: ah i think it's actually the _current_ here that's throwing me off... maybe this?
```suggestion
`b"EpochRewardsPartitionData"`) and the rewards distribution epoch number as a little-endian
```

though i don't recall now which we used in the code. i think it'd actually make more sense to seed off the epoch during which the rewards were earned?

Author: Yes, your wording suggestion seems better.

> though i don't recall now which we used in the code.

We used the distribution epoch number (`Bank::epoch()`).

> i think it'd actually make more sense to seed off the epoch during which the rewards were earned?

That may be more intuitive from a querying perspective, but I prefer to leave it as-is, since the distributions actually occur within the new epoch. I'll use the distribution epoch with EpochSchedule to find the first slot/block in the epoch.

record the necessary data in the ledger in some fashion. This could be a
transaction that gets added to the block, a special RewardType, or metadata that
gets stored on the node and then duplicated to long-term storage, like Bigtable.
In fact, it is probably worthwhile to pursue this alternative as well, since it
will enable finding rewards without access to snapshots or the running chain.

## New Terminology

None

## Detailed Design

When partitioned rewards are calculated in the runtime (currently during the
first block of the epoch), the runtime should populate a PDA that stores the
partition data needed to recreate the hasher that returns the partition index
for any address. This data comprises: the number of partitions, the parent
blockhash, and the hasher kind. More specifically:

```rust
// Version wrapper to allow future updates
// Variant serialized as little-endian unsigned 32-bit integer
enum EpochRewardsPartitionDataVersion {
    V0(PartitionData),
}

// Extensible list of kinds of hashers used to generate partitions
// Variant serialized as little-endian unsigned 32-bit integer
enum HasherKind {
    Sip13 = 0,

> }

Author: That works for me. @t-nelson, what's your opinion?

> }

Reviewer 2: fine by me. if we need to change it more than 4B times, that's on us

// Data about rewards partitions for a particular epoch
struct PartitionData {
    num_partitions: usize, // little-endian unsigned 64-bit integer
    parent_blockhash: Hash, // byte-array of length 32
    hasher_kind: HasherKind, // little-endian unsigned 32-bit integer
}
```

The address of this PDA should include some bytes to prevent griefing (eg.
`b"EpochRewardsPartitionData"`) and the rewards distribution epoch number as a

> little-endian u64. The owning program should be the Stake Program.

Author: Do you mean just state it more equivocally?

> little-endian u64. The owning program should be the Stake Program.

Author: I pushed a commit; let me know if that's not what you meant.
